exp/runtime/internal/controllers/extension_controller.go

```diff
@@ -21,74 +21,169 @@ import (
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/cluster-api/util/predicates"
 )
 
 // +kubebuilder:rbac:groups=runtime.cluster.x-k8s.io,resources=extension,verbs=get;list;watch;create;update;patch;delete
 
-// ExtensionReconciler reconciles a Extension object.
-type ExtensionReconciler struct {
+// Reconciler reconciles an Extension object.
+type Reconciler struct {
 	Client        client.Client
 	RuntimeClient runtimeclient.Client
-	Registry      registry.Registry
+
+	// WatchFilterValue is the label value used to filter events prior to reconciliation.
+	WatchFilterValue string
 }
 
-func (r *ExtensionReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
 	err := ctrl.NewControllerManagedBy(mgr).
-		For(&expv1.Extension{}).
+		For(&runtimev1.Extension{}).
+		Named("runtime/extension").
 		WithOptions(options).
+		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue)).
 		Complete(r)
-
 	if err != nil {
 		return errors.Wrap(err, "failed setting up with a controller manager")
 	}
 	return nil
 }
 
-func (r *ExtensionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
 	log := ctrl.LoggerFrom(ctx)
-	log.V(4).Info("Reconciling")
+	if !r.RuntimeClient.IsRegistryReady() {
+		err := r.syncRegistry(ctx)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "extensions controller not initialized")
+		}
+		// Note: there was an intermittent bug when allowing this to continue (likely due to patching the same object multiple times?)
+		return ctrl.Result{}, nil
+	}
 
-	ext := &expv1.Extension{}
-	err := r.Client.Get(ctx, req.NamespacedName, ext)
+	log.V(4).Info("Reconciling")
+	extension := &runtimev1.Extension{}
+	err := r.Client.Get(ctx, req.NamespacedName, extension)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			// Object not found, return.  Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
+			// Object not found, return.
 			return ctrl.Result{}, nil
 		}
 
 		// Error reading the object - requeue the request.
 		return ctrl.Result{}, err
 	}
 
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(extension, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// FIXME: could dedupe these patch calls. Ignoring for simplicity at this stage.
+	defer func() {
+		// Always attempt to patch the object and status after each reconciliation.
+		// Patch ObservedGeneration only if the reconciliation completed successfully
+		patchOpts := []patch.Option{}
+		if reterr == nil {
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{},
```

Comment:
We don't have .status.observedGeneration so I think we shouldn't try to set the field